File: CODE/T4013/T4013.cpp
#define nametask "T4013"
#include <bits/stdc++.h>
#define fi first
#define se second
#define mp make_pair
#define pb push_back
#define SZ(a) (int) a.size()
#define all(a) a.begin(), a.end()
#define FOR(i, a, b) for (int i = a; i <= b; i++)
#define FOD(i, b, a) for (int i = b; i >= a; i--)

using namespace std;
typedef long long ll;
typedef pair <int, int> pi;

ll A, B;
int d, K;
ll f[20][20];

vector <int> getdigit(ll x){
    vector <int> res;
    while (x > 0){
        res.pb(x % 10);
        x /= 10;
    }
    return res;
}

ll calc(int id, bool tight, int cntd, vector <int> digit){
    if (id < 0) return (cntd == K);
    if (!tight && f[id][cntd] != -1) return f[id][cntd];
    int k = (tight) ? digit[id] : 9;
    ll cnt = 0;
    FOR(i, 0, k){
        bool new_tight = (i == digit[id]) ? tight : 0;
        cnt += calc(id - 1, new_tight, cntd + (i == d), digit);
    }
    if (tight) return cnt;
    return (f[id][cntd] = cnt);
}

ll solve(){
    memset(f, -1, sizeof f);
    vector <int> digitA, digitB;
    digitA = getdigit(A - 1);
    digitB = getdigit(B);
    return calc(SZ(digitB) - 1, 1, 0, digitB) - calc(SZ(digitA) - 1, 1, 0, digitA);
}

signed main()
{
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    if (fopen(nametask".inp", "r")){
        freopen(nametask".inp", "r", stdin);
        freopen(nametask".out", "w", stdout);
    }
    cin >> A >> B >> d >> K;
    ll res = solve();
    cout << res;
    return 0;
}

File: CODE/T4013/T4013.py
import sys

sys.stdin = open("T4013.inp", "r")
sys.stdout = open("T4013.out", "w")

def get_digits(x):
    digits = []
    while x > 0:
        digits.append(x % 10)
        x //= 10
    return digits

def calc(idx, tight, count_d, digits, d, K, dp):
    if idx < 0:
        return 1 if count_d == K else 0
    
    if not tight and dp[idx][count_d] != -1:
        return dp[idx][count_d]
    
    max_digit = digits[idx] if tight else 9
    count = 0
    
    for i in range(max_digit + 1):
        new_tight = tight and (i == digits[idx])
        count += calc(idx - 1, new_tight, count_d + (i == d), digits, d, K, dp)
    
    if not tight:
        dp[idx][count_d] = count
    
    return count

def solve(A, B, d, K):
    digits_A = get_digits(A - 1)
    digits_B = get_digits(B)
    
    dp = [[-1] * 20 for _ in range(20)]
    count_B = calc(len(digits_B) - 1, True, 0, digits_B, d, K, dp)
    
    dp = [[-1] * 20 for _ in range(20)]
    count_A = calc(len(digits_A) - 1, True, 0, digits_A, d, K, dp)
    
    return count_B - count_A

A, B, d, K = map(int, input().split())
print(solve(A, B, d, K))
